8348828: Windows dll loading now resolves symlinks

benjaminp opened the pull request:
Deep in the bowels of `System.loadLibrary`, `File.getCanonicalPath()` is called on the target library file before it is passed to the system library loading APIs. In JDK-8003887, `File.getCanonicalPath` was altered to resolve symlinks on Windows. This had unintended consequences for passing a symlink to `System.loadLibrary` on Windows. The underlying Windows `LoadLibrary` API inspects the file name passed to it and adds a `.dll` extension if the it is not already present. Thus, if `System.loadLibrary` was given a symlink to a file and that file didn't have a `.dll` extension, `LoadLibrary` try to load nonexistent file and fail.

Fix this problem by appending a `.` to library paths `NativeLibraries`. This trailing dot inhibits `LoadLibrary`'s own appending behavior.

---------
- [x] I confirm that I make this contribution in accordance with the [OpenJDK Interim AI Policy](https://openjdk.org/legal/ai).

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- \[x\] Change must not contain extraneous whitespace
- \[x\] Commit message must refer to an issue
- \[ \] Change must be properly reviewed (2 reviews required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer), 1 [Author](https://openjdk.org/bylaws#author))

### Issue
 * [JDK-8348828](https://bugs.openjdk.org/browse/JDK-8348828): Windows dll loading now resolves symlinks (**Bug** - P4)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/24694/head:pull/24694` \
`$ git checkout pull/24694`

Update a local copy of the PR: \
`$ git checkout pull/24694` \
`$ git pull https://git.openjdk.org/jdk.git pull/24694/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 24694`

View PR using the GUI difftool: \
`$ git pr show -t 24694`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/24694.diff">https://git.openjdk.org/jdk/pull/24694.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/24694#issuecomment-2859437113)
</details>

benjaminp committed:
8348828: Windows dll loading now resolves symlinks

benjaminp commented:
@AlanBateman @bplb Hello! Any chance either of you could be induced to sponsor and review this? Also, any idea how to accelerate the OCA verification process?

AlanBateman commented:
This issue has the "oca" label so we cannot engage, sorry.

benjaminp commented:
> This issue has the "oca" label so we cannot engage, sorry.

Ah, sorry. I only asked because I saw a colleague of yours helping out with OCA verification over at https://github.com/openjdk/jdk/pull/23905#issuecomment-2762865417. Also, the [OCA FAQ](https://oca.opensource.oracle.com/?ojr=faq) suggests to "please reach out to project maintainers (e.g., through comments on your GitHub PR, Slack channels, or mailing lists)". Any breadcrumbs you can lay for me to escape this forest are appreciated!

gustavosimon commented:
Hello @benjaminp! I think you should wait for the OCA process — it's very fast. I'll just ping @dalibor, as he might be able to help.

AlanBateman commented:
Are you the submitter of JDK-8348828? Asking because the configuration described seems very unusual and sounds like something that changed the JDK image so the DLLs from the JDK bin directory are moved to somewhere else and sym link put in place instead. I think we also need to understand if there is a bug in File::getCanonicalPath before change the usage.

benjaminp commented:
> Are you the submitter of JDK-8348828?

Yes, I am.

> Asking because the configuration described seems very unusual and sounds like something that changed the JDK image so the DLLs from the JDK bin directory are moved to somewhere else and sym link put in place instead.

Correct, I explained a bit on the issue, but we have a system that replicates normal filesystem trees with symlink trees.

> I think we also need to understand if there is a bug in File::getCanonicalPath before change the usage.

I believe the problem is that `getCanonicalPath` is working too well and that symlink resolution of native libraries should be left to the system resolver.

AlanBateman commented:
Would it be possible to paste in here, or in the JBS issue, some examples of the path provided to LoadLibrary with some commentary on the sym links created on the file system.

bplb commented:
> > New-Item -Path .\jdk-24\bin\jimage.dll -ItemType SymbolicLink -Value .\rando

What if instead this were the following
```
> New-Item -Path .\jdk-24\bin\jimage.dll -ItemType SymbolicLink -Value .\rando.dll
```
where `.dll` is appended to the target?

> At this point, the path to load will be `$PWD\rando` due to symlink resolution.

Which is in fact correct, no?

> Entering the VM, `NativeLibraries.load` will eventually pass `$PWD\rando` to `LoadLibrary`. Since `.\rando` lacks a `.dll` extension on its face, `LoadLibrary` will add it, observe that `$PWD\rando.dll` doesn't exist and fail.

Indeed from the `LoadLibrary` documentation
```
If the string specifies a module name without a path and the file name extension is
omitted, the function appends the default library extension ".DLL" to the module name.
To prevent the function from appending ".DLL" to the module name, include a trailing
point character (.) in the module name string.
```
so a trailing `.` character on `rando` might equally resolve the problem?

benjaminp commented:
> > > New-Item -Path .\jdk-24\bin\jimage.dll -ItemType SymbolicLink -Value .\rando
> 
> What if instead this were the following
> 
> ```
> > New-Item -Path .\jdk-24\bin\jimage.dll -ItemType SymbolicLink -Value .\rando.dll
> ```
> 
> where `.dll` is appended to the target?

I expect that would work given the `LoadLibrary` documentation you've quoted.

> 
> > At this point, the path to load will be `$PWD\rando` due to symlink resolution.
> 
> Which is in fact correct, no?

Correct in that that's the library that should be loaded. `LoadLibrary` doesn't want to cooperate, though.

> 
> > Entering the VM, `NativeLibraries.load` will eventually pass `$PWD\rando` to `LoadLibrary`. Since `.\rando` lacks a `.dll` extension on its face, `LoadLibrary` will add it, observe that `$PWD\rando.dll` doesn't exist and fail.
> 
> Indeed from the `LoadLibrary` documentation
> 
> ```
> If the string specifies a module name without a path and the file name extension is
> omitted, the function appends the default library extension ".DLL" to the module name.
> To prevent the function from appending ".DLL" to the module name, include a trailing
> point character (.) in the module name string.
> ```
> 
> so a trailing `.` character on `rando` might equally resolve the problem?

Perhaps. In my usecase, files come into the system "anonymously"; files that make up the symlink tree are received before it's known what the layout of the tree will be. A link to a given file could turn out to be, e.g., `myaudio.mp3`, `MyClass.java`, or `jimage.dll`. I'm disinclined to append `.` or `.DLL` to every file in off chance that it turns out to be a library that Java wants to load.

bplb commented:
> > > At this point, the path to load will be `$PWD\rando` due to symlink resolution.
> > 
> > 
> > Which is in fact correct, no?
> 
> Correct in that that's the library that should be loaded. `LoadLibrary` doesn't want to cooperate, though.

So is it correct to say that the link resolution appears correct but `LoadLibrary`'s quirks cause the intended behavior to fail?

benjaminp commented:
> > > > At this point, the path to load will be `$PWD\rando` due to symlink resolution.
> > > 
> > > 
> > > Which is in fact correct, no?
> > 
> > 
> > Correct in that that's the library that should be loaded. `LoadLibrary` doesn't want to cooperate, though.
> 
> So is it correct to say that the link resolution appears correct but `LoadLibrary`'s quirks cause the intended behavior to fail?

Yes. The `getCanonicalPath` canonicalization indeed produces the file pointed to by the symlink. However, in the context of `NativeLibraries`, doing path canonicalization is at best unnecessary and at worst yields the linked bug.

bplb commented:
> > So is it correct to say that the link resolution appears correct but `LoadLibrary`'s quirks cause the intended behavior to fail?
> 
> Yes. The `getCanonicalPath` canonicalization indeed produces the file pointed to by the symlink.

All right, it is good to know that it is behaving in the desired manner.

> However, in the context of `NativeLibraries`, doing path canonicalization is at best unnecessary and at worst yields the linked bug.

You might be correct. We'll see what @AlanBateman and others have to say about it.

AlanBateman commented:
> You might be correct. We'll see what @AlanBateman and others have to say about it.

It's still puzzled as to why the DLLs have been moved from the JDK bin directory to some other location, and renamed so they don't have the ".dll" suffix. There most be some other product in the picture that we can't see. The quoted text from the Windows LoadLibrary documentation, where it appends the ".dll" suffix when not provided, is very useful as it  helps us understand why it fails.

As regards dropping the canonicalization then it would require thinking about the lock map used for mapping the library names to locks. It would need checking if it would break concurrent loading when using different names / file paths to the same DLL. There may be a route that involves adding a method to ClassLoaderHelper to post-process the path and the Windows version could add the "." when it doesn't have the ".dll" suffix.

benjaminp committed:
new approach: append . to file name on Windows

benjaminp commented:
> > You might be correct. We'll see what @AlanBateman and others have to say about it.
> 
> I'm still puzzled as to why the DLLs have been moved from the JDK bin directory to some other location, and renamed so they don't have the ".dll" suffix. There most be some other product in the picture that we can't see. The quoted text from the Windows LoadLibrary documentation, where it appends the ".dll" suffix when not provided, is very useful as it helps us understand why it fails.

Yes, the mystery product is my code. :) It makes a JDK image with a tree of symlinks to a data directory. The regular files in the data directory do not have file extensions.

> 
> As regards dropping the canonicalization then it would require thinking about the lock map used for mapping the library names to locks. It would need checking if it would break concurrent loading when using different names / file paths to the same DLL.

I can see that removing the canonicalization could break the lock map. (Though I see the current deduplication and locking scheme doesn't handle hardlinks. Probably it would have to be keyed by `FileKey` to do that.)

> There may be a route that involves adding a method to ClassLoaderHelper to post-process the path and the Windows version could add the "." when it doesn't have the ".dll" suffix.

I tried out this approach, and it solves my problem. I've now pushed the implementation.

bplb commented on src/java.base/windows/classes/jdk/internal/loader/ClassLoaderHelper.java:
```diff
@@ -103,4 +103,13 @@ static String[] parsePath(String ldPath) {
+    /**
+     * Add a dot to the end of the library file name ultimately passed to
+     * <code>LoadLibrary</code>. This prevents <code>LoadLibrary</code> from
+     * automatically infering a .DLL extension.
```
`infering` -> `inferring`

benjaminp committed:
fix spelling

bplb commented:
I built the JDK with the diff in 70e943c applied on Windows 11 and verified that it resolves the scenario presented in the issue description, namely that `jimage.dll` can be moved elsewhere, in this case `C:\Users\bpb\rando`, and a link created from `jimage.dll` to `rando` and with this config `java` and `jshell` can be run.

AlanBateman commented on src/java.base/share/classes/jdk/internal/loader/NativeLibraries.java:
```diff
@@ -121,7 +121,7 @@ public NativeLibrary loadLibrary(Class<?> fromClass, File file) {
                     return null;
                 }
-                name = file.getCanonicalPath();
+                name = file.getCanonicalPath() + ClassLoaderHelper.nativeLoaderFileNameSuffix();
```
For this proposal then the appending needs to conditional.

benjaminp replied:
Conditional on what?

AlanBateman replied:
The appending of the "." is specific to the case where there isn't a .dll suffix.

benjaminp replied:
> Is this documented in the Microsoft document for LoadLibrary? We don't want to rely on undocumented behavior as it could break at any time.

[LoadLibraryEx` documentation](https://learn.microsoft.com/en-us/windows/win32/api/libloaderapi/nf-libloaderapi-loadlibraryexa) states:
```
To prevent the function from appending ".DLL" to the module name, include a trailing point character (.) in the module name string.
```
which isn't incompatible with not adding `.dll` when the extension is already present.

Practically speaking, I imagine the behavior of a common operating system API available since at least Windows XP is completed ossified.

> 
> For the prototype, I think you can replace nativeLoaderFileNameSuffix() with mapToNativeLibraryName(String) so that other platforms can "return this". This would give the flexibility to prefix or do other mangling.

I've changed the implementation to do this.

> 
> That said, I'm not so sure that we should changing the JDK to work in this unusual environment. Are you able to name the product that is move-renaming the DLLs and replacing them with sym links? Just trying to get some sense as to whether this is something that many people might run into it.

I understand the narrow reproducer in the bug is weird but it was isolated from a generically useful system. As I explained in https://bugs.openjdk.org/browse/JDK-8348828?focusedId=14744766&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-14744766, this arises in a generic system my company offers for executing tasks in distributed fashion across many machines. We don't have any particular interest in DLLs or even JDK installations; we simply create a symlinks for all data required by tasks that our system is asked to execute.

This system worked fine with JDKS through Java 23 and all previous versions we've tried.

benjaminp committed:
Merge remote-tracking branch 'upstream/master' into nativelibraries-fix

benjaminp committed:
switch platform helper to be mapToNativeLibraryName

benjaminp commented:
/touch

liach commented:
Thanks for prompting for attention on the mailing list.

I just wonder if it would be better for us to handle this in `os::dll_load` in `os_windows.cpp` - this seems a better way to handle generic platform-specific peculiarities. This approach also means JNI and all other users would be updated too, but I don't see why JNI would handle this differently from class loader APIs.

benjaminp committed:
Merge remote-tracking branch 'origin/master' into nativelibraries-fix

benjaminp committed:
add dot in os::dll_load rather than NativeLibraries.java

benjaminp committed:
fix grammar

benjaminp committed:
fix compilation

benjaminp committed:
Merge branch 'master' into nativelibraries-fix

benjaminp committed:
use os::malloc

benjaminp committed:
add cast

dholmes-ora reviewed (changes requested):
I don't see that this is a hotspot issue and you can't just unconditionally add a dot to all passed in names - that will surely break any code that expected the .DLL to be appended!

As @AlanBateman has noted there are semantic question to address here. Library loading is tied to classloaders and the use of symlinks allows the same library to be loaded by different names in different classloaders - which could potentially wreak havoc I think if we execute the onLoad hooks multiple times.

I would like to see core-libs folk decide exactly what should happen for these different cases, and then figure out where is the best place to make that happen.

But initial view is that you should not be renaming the DLLs to not have a .dll extension - that just seems to be a self-inflicted wound.

benjaminp commented:
> I don't see that this is a hotspot issue and you can't just unconditionally add a dot to all passed in names - that will surely break any code that expected the .DLL to be appended!

I'm not sure such code previously would have worked. As far as I can tell, `System.loadLibrary` would fail before reaching `os::dll_load` if the string passed into `System.loadLibrary` could not be resolved to an extant file by `NativeLibraries`.

(Okay, I suppose that if `mylib` and `mylib.dll` both exist, `System.loadLibrary("mylib")` would actually load `mylib.dll`. But that actually seems bad?)

> 
> As @AlanBateman has noted there are semantic question to address here. Library loading is tied to classloaders and the use of symlinks allows the same library to be loaded by different names in different classloaders - which could potentially wreak havoc I think if we execute the onLoad hooks multiple times.

This change doesn't affect the resolution of symlinks now, though.

> 
> I would like to see core-libs folk decide exactly what should happen for these different cases, and then figure out where is the best place to make that happen.

liach's observation in https://github.com/openjdk/jdk/pull/24694#issuecomment-3377528055 that the JNI and `System.loadLibrary` should behave similarly likely means some hotspot change is needed.

> 
> But initial view is that you should not be renaming the DLLs to not have a .dll extension - that just seems to be a self-inflicted wound.

I agree that the reproducer in the bug is not something you should do for itself; it is a small self-contained way to demonstrate the regression. I tried to expand on what the broken application actually does a few more places if that helps such as here: https://bugs.openjdk.org/browse/JDK-8348828?focusedId=14744766&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-14744766.

dholmes-ora commented:
> I'm not sure such code previously would have worked. As far as I can tell, System.loadLibrary would fail before reaching os::dll_load if the string passed into System.loadLibrary could not be resolved to an extant file by NativeLibraries.

`os::dll_load` is not used only to implement `System.loadLibrary`.

benjaminp commented:
> > I'm not sure such code previously would have worked. As far as I can tell, System.loadLibrary would fail before reaching os::dll_load if the string passed into System.loadLibrary could not be resolved to an extant file by NativeLibraries.
> 
> `os::dll_load` is not used only to implement `System.loadLibrary`.

So, this dot logic should be lifted to `JVM_LoadLibrary`?

dholmes-ora commented:
> So, this dot logic should be lifted to JVM_LoadLibrary?

Possibly, but I still want to understand this change as it is now proposed. The whole thing has been about symlinks till now, so now what exactly is the problem and solution? You just want a way to load an arbitrarily-named file as if it were a dll?

benjaminp committed:
Merge branch 'master' into nativelibraries-fix

benjaminp committed:
revert dll_load fix

benjaminp committed:
add test showing loading symlinked libraries with various combinations

benjaminp commented:
> A better starting point for discussion would be a test that exercises System.loadLibrary with the a library name that locates a DLL through a sym link to a final target that doesn't have the .dll suffix.

To this end, I have updated the PR to add a test `test/jdk/java/lang/ClassLoader/loadLibrarySymlinks/LoadLibrarySymlinksTest.java`. This test exercises the 3 cases discussed in https://mail.openjdk.org/pipermail/core-libs-dev/2025-October/152992.html.

Unix platforms pass this test as expected. Since I reverted all of my proposed fixes, Windows fails handling the `test2.dll -> barename` case:
```
java.lang.UnsatisfiedLinkError: D:\\a\\jdk\\jdk\\build\\run-test-prebuilt\\test-support\\jtreg_test_jdk_tier1_part1\\scratch\\barename: Can't find dependent libraries
        at java.base/jdk.internal.loader.NativeLibraries.load(Native Method)
        at java.base/jdk.internal.loader.NativeLibraries$NativeLibraryImpl.open(NativeLibraries.java:326)
        at java.base/jdk.internal.loader.NativeLibraries.loadLibrary(NativeLibraries.java:187)
        at java.base/jdk.internal.loader.NativeLibraries.loadLibrary(NativeLibraries.java:129)
        at java.base/java.lang.ClassLoader.loadLibrary(ClassLoader.java:2282)
        at java.base/java.lang.Runtime.load0(Runtime.java:767)
        at java.base/java.lang.System.load(System.java:1646)
        at LoadLibrarySymlinksTest.main(LoadLibrarySymlinksTest.java:48)
        at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:104)
        at java.base/java.lang.reflect.Method.invoke(Method.java:565)
        at com.sun.javatest.regtest.agent.MainActionHelper$AgentVMRunnable.run(MainActionHelper.java:335)
        at java.base/java.lang.Thread.run(Thread.java:1474)

JavaTest Message: Test threw exception: java.lang.UnsatisfiedLinkError
```

I hope this aids in understanding the issue and review my proposed fix(es).

AlanBateman commented:
> I hope this aids in understanding the issue and review my proposed fix(es).

I think the issue is understood. The original bug report was a bit confusing as it involved replacing the DLLs in the JDK bin directory with sym links. I think we are past that now and the discussion has moved onto sym links with the .dll suffix that are linked to files that don't have the .dll suffix.

We don't have an insight as to why the Windows APIs requires "opt-in" for this.  It may help to thwart attempts to evade scanners, maybe it makes it harder for smugglers, or maybe there are other reasons. I've asked our security team for some input on the topic before forming an opinion as to whether we should do anything.

benjaminp committed:
Merge branch 'master' into nativelibraries-fix

benjaminp commented:
Any indications from security?

benjaminp committed:
Merge branch 'master' into nativelibraries-fix

benjaminp committed:
update copyright years of test files

benjaminp committed:
Merge branch 'master' into nativelibraries-fix

benjaminp committed:
append '.' to library names on Windows so that LoadLibrary doesn't mangle them

benjaminp commented:
/label remove hotspot-runtime

benjaminp commented:
/template append

benjaminp commented:
I've reinstanted the "core-libs" solution, which is to append "." on Windows in `NativeLibraries`. The test demonstrates that the regression is fixed.

benjaminp commented:
I wonder if security has come back with any thoughts?

benjaminp committed:
Merge branch 'master' into nativelibraries-fix

benjaminp commented:
Any messages drift back from the security divisions?

benjaminp committed:
Merge branch 'master' into nativelibraries-fix

benjaminp committed:
make aix copyright happy
